pipeline.ml

```diff
@@ -12,118 +12,144 @@ let read_fpath p = Bos.OS.File.read p |> Rresult.R.error_msg_to_invalid_arg
-  let src = Git.fetch (Current.map Github.Api.Commit.id head) in
   let dockerfile =
     let+ base = Docker.pull ~schedule:weekly "ocaml/opam2" in
     dockerfile ~base
   in
-  let image = Docker.build ~pull:false ~dockerfile (`Git src) in
-  let docker_cpuset_cpus =
-    match docker_cpu with
-    | Some i -> [ "--cpuset-cpus"; string_of_int i ]
-    | None -> []
-  in
-  let docker_cpuset_mems =
-    match docker_numa_node with
-    | Some i -> [ "--cpuset-mems"; string_of_int i ]
-    | None -> []
-  in
-  let tmpfs =
-    match docker_numa_node with
-    | Some i ->
-        [
-          "--tmpfs";
-          Fmt.str "/dev/shm:rw,noexec,nosuid,size=%dG,mpol=bind:%d"
-            docker_shm_size i;
-        ]
-    | None ->
-        [
-          "--tmpfs";
-          Fmt.str "/dev/shm:rw,noexec,nosuid,size=%dG" docker_shm_size;
-        ]
-  in
-  let s =
-    let run_args =
-      [
-        "--security-opt";
-        "seccomp=./aslr_seccomp.json";
-        "--mount";
-        Fmt.str "type=bind,source=%a,target=%a" Fpath.pp tmp_source Fpath.pp
-          tmp_target;
-      ]
-      @ tmpfs
-      @ docker_cpuset_cpus
-      @ docker_cpuset_mems
-    in
-    let+ () =
-      Docker.run ~run_args image
-        ~args:
-          [
-            "/usr/bin/setarch";
-            "x86_64";
-            "--addr-no-randomize";
-            "_build/default/bench/db_bench.exe";
-            "--bench";
-            "index";
-            "--json";
-            Fmt.str "%a" Fpath.pp tmp_target;
-          ]
-    in
-    (* Conditionally move the results to ?output_file *)
-    let results_path =
-      match output_file with
-      | Some path ->
-          Bos.OS.Path.move tmp_source path |> Rresult.R.error_msg_to_invalid_arg;
-          path
-      | None -> tmp_source
-    in
-    (* No need to read JSON if we're not publishing the results anywhere *)
-    match slack_path with
-    | Some p -> Some (p, read_fpath results_path)
-    | None -> None
-  in
-  s
-  |> Current.option_map (fun p ->
-         Current.component "post"
-         |> let** path, _ = p in
-            let channel = read_channel_uri path in
-            Slack.post channel ~key:"output" (Current.map snd p))
-  |> Current.ignore_value
+  Github.App.installations app
+  |> Current.list_iter ~pp:Github.Installation.pp @@ fun installation ->
+     let repos = Github.Installation.repositories installation in
+     repos
+     |> Current.list_iter ~pp:Github.Api.Repo.pp @@ fun repo ->
+        Github.Api.Repo.ci_refs repo
+        |> Current.list_iter ~pp:Github.Api.Commit.pp @@ fun head ->
+           let src = Git.fetch (Current.map Github.Api.Commit.id head) in
+           let image = Docker.build ~pool ~pull:false ~dockerfile (`Git src) in
+           let docker_cpuset_cpus =
+             match docker_cpu with
+             | Some i -> [ "--cpuset-cpus"; string_of_int i ]
+             | None -> []
+           in
+           let docker_cpuset_mems =
```

Comment:
Would it be possible to leave these helper functions one level up? They don't seem to change the value in each iteration and would make it a bit easier to see what's new

Reply:
Yep will do!